docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+   This will potentially allow the class to match interfaces that it did not previously and may affect the behaviour of code that assumes specific semantics from classes that match the interface.
+   It will also cause compile failures for any code that uses implementation inheritance to reuse the code from that class and defines methods or fields of the same name.
+ - Adding any top-level public field, method, or type.
+   This will break any file that imports the module into its top-level scope and defines something with the same name (or imports something else into its top-level scope that defines something with the same name).
+
+In most packages, it will be important to make this kind of breaking change periodically.
+The simplest way for a package author to do this is simply to bump the library version and require everyone to fix any breakages in their code.
+This is not desirable because the only way that this allows *graceful deprecation*, which provides a window in which old and new versions of the APIs exist, is to ship two versions of the library.
+Shipping two versions of the library causes several problems, most notably:
+
+ - Security fixes must be back-ported to the old code.
+   This is somewhat tractable when you have two versions but if this is needed for every API change then it can rapidly become dozens of versions.
+ - If a project uses package A and also package B that uses package A, then it is possible to end up with a dependency on two incompatible versions of package A.
+
+API Versioning
+--------------
+
+In Verona, we differentiate between the package (implementation) version and the API version.
+A package version is a simple monotonic counter of package releases.
+For each release, the package supports a set of API versions.
+API versions use semantic versioning, where the major version is incremented to indicate the removal of types or methods and the minor version to indicate the addition of new classes or methods.
+Both indicate potentially breaking changes for Verona consumers but it should be possible for a package to support all minor versions of an API with minimal effort by the package author.
+There is also one special API version: `unstable`.
+This represents any work-in-progress changes since the last release and exists so that developers can import the current snapshot of a library to test whether in-development changes are likely to cause problems.
+
+By making the API version distinct from the implementation version, we allow:
+
+ - A program to consume two different API versions of the same package (while consuming a single implementation)
+ - A package to deprecate and remove APIs gradually over time without needing two versions of the code.
+ - A package to add new APIs without breaking any consumers.
+
+When a package is imported, the string describing the package should include the API version, for example:
+
+```verona
+import "some/package@2.0" as p;
```

Comment:
Would be nice to have a more complete example, like including the version annotation seen in the corner cases section below.